packages/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/ImagePickerDelegate.java

```diff
@@ -0,0 +1,254 @@
+import java.util.UUID;
+
+public class ImagePickerDelegate
+    implements PluginRegistry.ActivityResultListener,
+        PluginRegistry.RequestPermissionsResultListener {
+  private static final int REQUEST_CODE_CHOOSE_FROM_GALLERY = 2342;
+  private static final int REQUEST_CODE_TAKE_WITH_CAMERA = 2343;
+
+  private static final int REQUEST_EXTERNAL_STORAGE_PERMISSION = 2344;
+  private static final int REQUEST_CAMERA_PERMISSION = 2345;
+
+  private final Activity activity;
+  private final File externalFilesDirectory;
+  private final ImageResizer imageResizer;
+  private final String providerName;
+
+  private Uri pendingCameraImageUri;
+  private MethodChannel.Result pendingResult;
+  private MethodCall methodCall;
+
+  public ImagePickerDelegate(
+      Activity activity, File externalFilesDirectory, ImageResizer imageResizer) {
+    this.activity = activity;
+    this.externalFilesDirectory = externalFilesDirectory;
+    this.imageResizer = imageResizer;
+    this.providerName = activity.getPackageName() + ".flutter.image_provider";
+  }
+
+  public void chooseImageFromGallery(MethodCall methodCall, MethodChannel.Result result) {
+    if (!setPendingMethodCallAndResult(methodCall, result)) {
+      return;
+    }
+
+    boolean hasExternalStoragePermission =
+        ActivityCompat.checkSelfPermission(activity, Manifest.permission.READ_EXTERNAL_STORAGE)
+            == PackageManager.PERMISSION_GRANTED;
+
+    if (!hasExternalStoragePermission) {
+      requestExternalStoragePermission();
+      return;
+    }
+
+    launchPickImageFromGalleryIntent();
+  }
+
+  private void requestExternalStoragePermission() {
+    ActivityCompat.requestPermissions(
+        activity,
+        new String[] {Manifest.permission.READ_EXTERNAL_STORAGE},
+        REQUEST_EXTERNAL_STORAGE_PERMISSION);
+  }
+
+  private void launchPickImageFromGalleryIntent() {
+    Intent pickImageIntent = new Intent(Intent.ACTION_GET_CONTENT);
+    pickImageIntent.setType("image/*");
+
+    activity.startActivityForResult(pickImageIntent, REQUEST_CODE_CHOOSE_FROM_GALLERY);
+  }
+
+  public void takeImageWithCamera(MethodCall methodCall, MethodChannel.Result result) {
+    if (!setPendingMethodCallAndResult(methodCall, result)) {
+      return;
+    }
+
+    boolean hasCameraPermission =
+        ActivityCompat.checkSelfPermission(activity, Manifest.permission.CAMERA)
+            == PackageManager.PERMISSION_GRANTED;
+
+    if (!hasCameraPermission) {
+      requestCameraPermission();
+      return;
+    }
+
+    launchTakeImageWithCameraIntent();
+  }
+
+  private void requestCameraPermission() {
+    ActivityCompat.requestPermissions(
+        activity, new String[] {Manifest.permission.CAMERA}, REQUEST_CAMERA_PERMISSION);
+  }
+
+  private void launchTakeImageWithCameraIntent() {
+    Intent intent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
+    boolean canTakePhotos = intent.resolveActivity(activity.getPackageManager()) != null;
+
+    if (canTakePhotos) {
```

Comment:
nit: I would find this easier to understand if you turn it arround:

```java
if(!canTakePhotos) {
  finishWithError(...);
  return;
}

// take photo
```

This also applies to code further down.